### ISSUE TITLE
add 3 samples to Windows Universal Samples in GitHub

### ISSUE THREAD

**Issue by optikos** [MicrosoftDocs/windows-uwp#406]
The 3 samples {Bookstore1, QuizGame, Bookstore2} that are linked to herein should also appear in Windows Universal Samples on GitHub.  Please add them for both C# and C++/CX (so that the C++/WinRT variant is then added there too).

https://github.com/Microsoft/Windows-universal-samples

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 918404cb-28b0-1624-da04-c992bfca4395
* Version Independent ID: 20df39b5-a4db-3065-a172-5a90d27909fb
* Content: [Data binding in depth - UWP app developer](https://docs.microsoft.com/en-us/windows/uwp/data-binding/data-binding-in-depth)
* Content Source: [windows-apps-src/data-binding/data-binding-in-depth.md](https://github.com/MicrosoftDocs/windows-uwp/blob/docs/windows-apps-src/data-binding/data-binding-in-depth.md)
* Service: **unspecified**
* GitHub Login: @mcleblanc
* Microsoft Alias: **markl**

**Comment by stevewhims** [comment]
Hi, @optikos, thanks for raising this issue, I know you filed it a long time ago, and I'm sorry for the really delayed response here.

I initially wrote Bookstore1 and 2 for some content around porting from Silverlight and from Windows 8.x apps to Windows 10 apps. The Windows-universal-samples are meant to show off rather focused feature areas based on related APIs. Bookstore1 and 2 do have a decent center of gravity around data binding, but that's not all. Suffice to say, I feel that they're different enough that they oughtn't to be included with the feature area samples. A similar thing is true of QuizGame. QuizGame was meant to cover several different feature areas in one sample app, and show the synergy between them. It was always meant to be a different, broader-in-scope, kind of sample app from the feature area samples and so it was by design not included with the feature area samples.

Adding them to the feature area samples won't automatically mean that they're ported to new language projections, either. However, we do drive that kind of porting by customer demand, so your ask here will certainly be noted, and taken into account.

To the extent that we have no immediate plans to port these three samples to C++/WinRT (and we wouldn't port them to C++/CX at this stage, since that's now been superseded by C++/WinRT), I think it's best that I close this issue for now. However, Bookstore does live on in a minor form in the context of C++/WinRT here: https://docs.microsoft.com/windows/uwp/cpp-and-winrt-apis/binding-property.

Thanks again! And, please feel free to re-open and/or comment further here if you feel I've closed the issue prematurely.

~Steve

**Comment by optikos** [comment]
(sigh)  Microsoft's commitment to C++/WinRT is not appreciably more (yet) than Microsoft's former commitment to C++/CX.  C# is never the poor orphan step child.  C++ developers don't ask for every educational material in C# to be ported to C++/WinRT (or back in the day, to C++/CX or even further back, to C++/CLI), but the educational material for C++/* should minimally cover all the bases.  Kenny Kerr's team tends to cherry pick their examples a little bit.  I'd like to see examples that go one increment beyond what Kenny Kerr's team cherry picks.  (Obviously no one wants to air out dirty laundry to showcase an area where C++/WinRT is lackluster or worse cannot do at all, but Microsoft shouldn't let that duty fall to the public to wander aimlessly trying to figure it out.)

Specifically there are areas where C#'s "as" clause comes in mighty handy that I have not fully understood in C++/WinRT.  C++/CX has slightly fewer of these lost-in-translation areas.

I don't mind you closing this one individual issue, but **I do very much mind Microsoft as a collective failing** to inventory the 100 or so commonplace programming idioms in C# that need to have analogous idioms in C++/WinRT (or in an earlier era, C++/CX or C++/CLI).  Say, 88 of the 100 are well-covered by examples provided by Kenny Kerr's team.  Say, 5 of the remaining 12 are able to be guessed by adequately-experienced developers, even if they have never seen that topic before in textbook-esque examples.  That leaves, say, 7 out of the 100 C# idioms that C# developers take for granted that C++/WinRT developers are left pondering and pondering and pondering what clever C++-ish trick to do because neither C++ nor C++/WinRT have a direct equivalent.  Something in the vicinity of the "as" clause and typecasting in C# is a recurring pain for me in C++/WinRT development.  Instead of learning some idiom (as I would very much like to do), I radically redesign the UI to no longer need that "as" clause if it were C# instead of C++/WinRT.  This is incredibly inefficient and disheartening.

We don't need the Bookstore app example for C++/WinRT per se, but we need some way of teaching the, say, 7 C# idioms that Kenny Kerr's team haven't cherry picked in their examples (likely for a reason) of which C++ lacks a direct equivalent and of which C++/WinRT lacks a direct equivalent (that works in *all* translated-from-C#-UI cases instead of cherry-picked cases).

**Comment by stevewhims** [comment]
@optikos, regarding the "examples cherry-picked by Kenny Kerr's team", are you referring to the code examples that appear on GitHub, and Kenny's blog? I'm hoping to understand *where* the gaps are, and *what* they are.

If there's anything I can add to the official C++/WinRT docs, whether [conceptual](https://docs.microsoft.com/windows/uwp/cpp-and-winrt-apis/index) or [API reference](https://docs.microsoft.com/uwp/cpp-ref-for-winrt/winrt), then it would be great if you could leave a GitHub issue on a topic in one of those content sets, and let me know what's missing. If you do so, then I'll automatically see it. Those content sets are the ones that are currently officially supported when it comes to C++/WinRT documentation.

There's documentation for **as** and **try_as** in [Instantiating and returning projected types and interfaces](https://docs.microsoft.com/windows/uwp/cpp-and-winrt-apis/consume-apis#instantiating-and-returning-projected-types-and-interfaces), the [as function](https://docs.microsoft.com/uwp/cpp-ref-for-winrt/windows-foundation-iunknown#iunknownas-function), and the [try_as function](https://docs.microsoft.com/uwp/cpp-ref-for-winrt/windows-foundation-iunknown#iunknowntryas-function). As those docs say, the notion of **as** is essentially that of **IUnknown::QueryInterface**. You can think of it as a cast (similar to C#'s as), but it's probably best to think of it as performing a QueryInterface, because that is actually how it's implemented. Please let me know whether this info helps and, if not, what's missing from the info. For example, are you needing to cast between projected and implementation types? For that, a topic such as [winrt::get_self](https://docs.microsoft.com/uwp/cpp-ref-for-winrt/get-self) would be helpful. That topic in turn links to conceptual content around casting in both directions, as well as in-depth docs around the distinction between, and purpose of, implementation types and projected types.

~Steve

**Comment by optikos** [comment]
(sigh) Yes, I am aware of try_as.  I don't know what it is that I don't know (or else I might perhaps have the insight to provide the solution to my own gap of knowledge):  a chicken & egg problem.  There is something in the vicinity of try_as that I don't get.  The cherry-picked examples from Kenny Kerr's team seem to pick only the low-hanging fruit.  They don't seem to take something substantial like this Bookstore app in C# and really dig deep to transliterate every C# idiom utilized in this Bookstore app to C++/WinRT.  That is why I requested that this Bookstore app would be a wonderful demonstration of transliterating C# idioms directly or indirectly into whatever the mapping to C++/WinRT would be, without skipping any UI-functionality portion of this Bookstore app.

I learn new tech best when I see a direct before-&-after comparison of the prior tech to the new tech for the same topic.  So far, both your closing of this issue and the cherry-picking of low-hanging-fruit examples in GitHub by Kenny Kerr's team are obstructing my ability to learn the remaining few of the 100-ish C++/WinRT idioms that are the analogue of the 100-ish C# idioms that C# programmers take for granted.

I am rather frustrated by road-blockage after road-blockage after road-blockage from various portions of Microsoft on revealing/teaching this information.  I don't know what I don't know.  It is in the vicinity of using try_as to port the C# Bookstore app to C++/WinRT.  It is also somewhere in the vicinity of C++/WinRT types for UWP .winMD types not quite being full-fledged C++ types.  My lack of understanding falls somewhere in the vicinity of 1) C#'s as, 2) C++/WinRT's try_as, 3) the UWP DLL ASP "types", 4) C++/WinRT's disconnect with full-fledged C++ types, and 5) language-level typecasting being a no-no on this topic in C++/WinRT but being at the heart of C#'s as mechanism.  The C# community doesn't suffer such road-blockages on all fronts from all departments at Microsoft; Microsoft is so incredibly generous with C# educational material.  But C++/CLI then C++/CX and now C++/WinRT repeatedly does suffer from such road-blockages.

If even one of the various departments (and this isn't the first/only GitHub issue that requests the porting of existing C# or C++/CX app to C++/WinRT outside of Kenny Kerr's cherry-picking-examples team) would expend as much effort in actually porting an entire already-existing substantial C# example/sample app to C++/WinRT as they expend in rejecting my requests, we would have had a not-cherry-picked C++/WinRT example/sample app by now.  But instead of actually learning what I don't understand regarding C++/WinRT's try_as (or learning where C++/WinRT has an overt lack of feature that I keep butting my head into), I learn to be the best GitHub reply-commentator to rejected GitHub issues.  (Sharpen the saw.  What what one does 10,000 hours or 10,000 times, one becomes an expert at.  That's the popular adage, right?)  Which would Microsoft rather I be:

- a C++/WinRT programmer who gets past my try_as obstacle via one of these request to port an already-existing C# example/sample app to C++/WinRT, or

- forever be a professional reply-commentator to rejected reasonable-request GitHub issues.

It gets quite old having multiple Microsoft departments reject my GitHub issue requests to port an already-existing C# example/sample app to C++/WinRT so that I can compare not-cherry-picked substantial set of C# idioms mapped into analogous C++/WinRT idioms.  But I am a patient man.  Someday I will find the Microsoft employee who works for the right manager who doesn't reject my reasonable-request GitHub issues that request the C#-to-C++/WinRT port of a substantial example/sample app.

Normally, in the face of repeated road-blocks for learning a topic directly from the source, I would just give up on direct-from-the-source learning and go buy a multihundred-page textbook from which to absorb the whole topic.  But for C++/WinRT, there is absolutely no book available on Amazon.com.  So I keep coming back here to submit request-to-port-sample-apps GitHub issues (and radically redesigning my UIs to bypass the need for try_as).  Perhaps the reason for this total lack of books on C++/WinRT is that the authors who try to write such a book for C++/WinRT are having the same roadblocks as I am.

**Comment by stevewhims** [comment]
I hear you, @optikos, and I'll see what I can do regarding providing C++/WinRT code examples for this topic. It's definitely on my list, and your feedback helps me to prioritize it.

~Steve

**Comment by optikos** [comment]
Thank you.  Even *listening* to me this far is further than I have gotten on other similar GitHub issues on this topic.

**Comment by stevewhims** [comment]
Hi, @optikos, I've made some progress porting Bookstore2 to C++/WinRT. On the to which this issue is attached, the Bookstore2 link (under the heading "Sample apps that demonstrate {Binding}") now downloads a zip that contains both cs and cppwinrt folders, each folder containing a functionally identical app in the respective language projection. One caveat is that I'm still getting the C++/WinRT version code-reviewed. But I didn't want that to block me publishing this preliminary version in case it unblocks you in any way. I'll post a follow-up here when this version's code-reviewed and final. Thanks for your patience! ~Steve

**Comment by stevewhims** [comment]
One last update, here. I've made the final changes to the C++/WinRT version of Bookstore2 after the code-review pass. I hope it's useful. ~Steve
